### PR TITLE
[inductor] Add type hints to functions in decompositions.py

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131780

Summary: ATT

Test Plan: lintrunner

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -109,8 +108,8 @@
 remove_decompositions(decompositions, decomps_to_exclude)
 
 
-def register_decomposition(ops):
-    for op in [ops] if callable(ops) else ops:
+def register_decomposition(ops: List[torch._ops.OperatorBase]) -> Callable[..., Any]:
```
Comment (eellison) [inline torch/_inductor/decomposition.py:111]: This should be `Union[OpOverloadPacket, OperatorBase]`

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -119,24 +118,33 @@ def register_decomposition(ops):
...
     return
 
 
 @register_decomposition([aten.clamp])
 @pw_cast_for_opmath
-def clamp(x, min=None, max=None):
+def clamp(
+    x: torch.Tensor,
+    min: Optional[Union[int, float]] = None,
+    max: Optional[Union[int, float]] = None,
```
Comment (eellison) [inline torch/_inductor/decomposition.py:146]: Can we use `torch.types.Number` here ?

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -145,7 +153,11 @@ def clamp(x, min=None, max=None):
 
 
 @register_decomposition([aten.full])
-def full(size, fill_value, **kwargs):
+def full(
+    size: List[Union[int, torch.SymInt]],
+    fill_value: Union[int, float],
```
Comment (eellison) [inline torch/_inductor/decomposition.py:158]: ditto

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -282,10 +310,13 @@ def mm(self, input2):
 # - Normalize cat calls, so that legacy empty 1-D tensors are removed (NB: we
 #   don't remove ALL empty tensors, only the naughty ones)
 @register_decomposition([aten.cat.default])
-def cat(tensors, dim=0):
+def cat(
+    tensors: torch.Tensor,
```
Comment (eellison) [inline torch/_inductor/decomposition.py:314]: this should be list of tensors right ?

File: torch/_inductor/decomposition.py [hunk 5]
```diff
@@ -342,7 +373,12 @@ def angle(x):
 
 
 @register_decomposition([aten.add])
-def add(x, y, *, alpha=None):
+def add(
+    x: torch.Tensor,
+    y: torch.Tensor,
+    *,
+    alpha: Optional[float] = None,
```
Comment (eellison) [inline torch/_inductor/decomposition.py:380]: alpha `float` -> Number

File: torch/_inductor/decomposition.py [hunk 6]
```diff
@@ -578,21 +678,35 @@ def grid_sampler_2d(
 
 
 @register_decomposition(aten._foreach_addcmul.Scalar)
-def _foreach_addcmul_scalar(self, left_tensors, right_tensors, scalar=1):
+def _foreach_addcmul_scalar(
+    self: torch.Tensor,
```
Comment (eellison) [inline torch/_inductor/decomposition.py:682]: self is a `List[Tensor]`

File: torch/_inductor/decomposition.py [hunk 7]
```diff
@@ -578,21 +678,35 @@ def grid_sampler_2d(
...
+) -> List[torch.Tensor]:
     return aten._foreach_add.List(
         self, aten._foreach_mul.List(left_tensors, right_tensors), alpha=scalar
     )
 
 
 @register_decomposition(aten._foreach_addcdiv.Scalar)
-def _foreach_addcdiv_scalar(self, left_tensors, right_tensors, scalar=1):
+def _foreach_addcdiv_scalar(
+    self: torch.Tensor,
```
Comment (eellison) [inline torch/_inductor/decomposition.py:694]: ditto

File: torch/_inductor/decomposition.py [hunk 8]
```diff
@@ -578,21 +678,35 @@ def grid_sampler_2d(
...
         self, aten._foreach_div.List(left_tensors, right_tensors), alpha=scalar
     )
 
 
 @register_decomposition(aten._foreach_lerp.Scalar)
-def _foreach_lerp_scalar(start_tensors, end_tensors, weight):
+def _foreach_lerp_scalar(
+    start_tensors: List[torch.Tensor],
+    end_tensors: List[torch.Tensor],
+    weight: float,
```
Comment (eellison) [inline torch/_inductor/decomposition.py:708]: weight -> Number

File: torch/_inductor/decomposition.py [hunk 9]
```diff
@@ -634,19 +748,23 @@ def miopen_batch_norm(
 
 
 @functools.lru_cache(None)
-def fast_random_decomps():
+def fast_random_decomps():  # type: ignore[no-untyped-def]
```
Comment (eellison) [inline torch/_inductor/decomposition.py:751]: `-> Dict[OpOverload, Callable[[Any], Any]` ?

File: torch/_inductor/decomposition.py [hunk 10]
```diff
@@ -634,19 +748,23 @@ def miopen_batch_norm(
...
     return {**decompositions, **extra_random_decomps}
 
 
-def select_decomp_table():
+def select_decomp_table():  # type: ignore[no-untyped-def]
```
Comment (eellison) [inline torch/_inductor/decomposition.py:755]: same

**Review by eellison — APPROVED** [review]
looks good! a few comments

**Review by eellison — APPROVED** [review]
looks good! a few comments

File: torch/_inductor/decomposition.py (hunk 4 above)
Comment (aakhundov) [reply torch/_inductor/decomposition.py:314]: Yep, great catch!

Comment (aakhundov) [reply torch/_inductor/decomposition.py:314]: Wondering how did mypy not complain...

File: torch/_inductor/decomposition.py (hunk 9 above)
Comment (aakhundov) [reply torch/_inductor/decomposition.py:751]: So I've tried this and a few external call sites have started breaking. Let me try again.

File: torch/_inductor/decomposition.py (hunk 4 above)
Comment (eellison) [reply torch/_inductor/decomposition.py:314]: We probably dont invoke this function directly much.. only through decomps

File: torch/_inductor/decomposition.py (hunk 9 above)
Comment (aakhundov) [reply torch/_inductor/decomposition.py]: So the best I could do is `Dict[Any, Callable[..., Any]]` without ripples in other files. Left a TODO.
